### ISSUE TITLE
如何获取学生在不同知识点上的掌握程度，即论文Figure.4中的CKS

### ISSUE THREAD

**Issue by ZhijieXiong** [lilstrawberry/MIKT#6]
<img width="466" height="480" alt="Image" src="https://github.com/user-attachments/assets/111a7882-0e89-461d-8ac6-2a85b1f789b2" />
请问代码中该如何获取论文中Figure.4中所示的CKS，即学生在各知识点上的掌握程度？因为我想使用MIKT作为模拟器，所以需要通过MIKT获取学生知识状态。
如能解惑，万分感激！

**Comment by lilstrawberry** [comment]
model.py第183行得到的就是对各个知识点的知识状态（在当前时间步下 维度为 batch skill d），看你后续怎么处理了

**Comment by ZhijieXiong** [comment]
<img width="1031" height="76" alt="Image" src="https://github.com/user-attachments/assets/01430bcd-180a-477d-8f05-3b4aaf073cc9" />
感谢回答，我debug了一下，看到skill_state是(bs, num_concept, dim)的向量，我是想获取一个shape为(bs, num_concept)，且值为0～1之间（或者一个能量化的范围）的向量，表示学生在每个知识点上的掌握程度。请问这个你是如何获取的，是用这个skill_state去和concept embedding做点积操作吗？

**Comment by ZhijieXiong** [comment]
就是论文图4中CKS输出学生在3个知识点上的掌握程度，这个我不知道该怎么获取

**Comment by lilstrawberry** [comment]
论文输出的不是在各个知识点的掌握 而且对各个知识点的关注（attention）。
按你的需求，你可以用当前的题目 batch d（扩充成batch skill d） 和 skill_state (batch skill d)  concat一下，然后过一个mlp得到[0，1]的分数

**Comment by ZhijieXiong** [comment]
好的，我明白了，非常感谢

**Comment by lilstrawberry** [comment]
如果你要得到比较客观的学生随时间对知识点分数的变化，我觉得用skill_state和concept embedding做一个mlp比较好（这个角度可能更常见，不过本文可能更符合下面的做法）。

如果你要得到学生聚焦于当前问题的知识点分数（也就是透过当前问题能推出学生的知识点分数），我刚刚说的那个可能好一些（这个角度更重视回答问题）。

（反正两个你都可以试试 哪个可解释性好就用哪种）

**Comment by ZhijieXiong** [comment]
非常感谢耐心解答。是这样的，我想找一个KT模型做LPR（学习路径推荐）的环境模拟器，该KT模型不仅需要能预测做题结果，还需要返回每个时刻学生在所有知识点上的掌握｜理解程度，以供推荐算法使用。大多数KT模型都不提供获取知识状态的方法，例如qDKT，我是通过将question emb置零，然后和concept emb拼接后送入预测器，才能得到知识状态。我感觉目前的KT模型MIKT是可解释性和性能兼顾最好的（我做了一个[榜单](https://zhijiexiong.github.io/sub-page/pyedmine/rankingList.html)，MIKT和ABQR是目前性能最好的两个模型 ），所以想使用MIKT做模拟器。
你说的两种方法，都需要一个MLP，可是这个MLP如何参与进训练，我不是很清楚。我想的是，用skill_state和concept embedding做某些无训练（例如内积）的操作得到一个标量值，来标准知识状态，你觉得这个是否合理？

**Comment by lilstrawberry** [comment]
那我觉得你修改一下MIKT的预测层，比如通过所有相关skill分数加权减去题目难度得到答案做训练（当然如果这么做 记得加上DKS也就是LSTM得到的知识状态（也可以导出一个总体分数），因为MIKT的性能需要它）

**Comment by lilstrawberry** [comment]
我觉得在实验下 不过多牺牲MIKT性能前提下合理的魔改都可以

**Comment by ZhijieXiong** [comment]
好的，明白了。多谢

**Comment by lilstrawberry** [comment]
> MIKT

榜单能不能把ReKT加一下，有点好奇它的水平🤣
paper: https://dl.acm.org/doi/abs/10.1145/3664647.3681205
code: https://github.com/lilstrawberry/ReKT

如果还方便的话，测测 https://github.com/lilstrawberry/DiffuQKT 的水平🤣🤣🤣

**Comment by ZhijieXiong** [comment]
ReKT我后面有空复现一下，不过我现在没做KT了，做LPR也是自己感兴趣，空闲时搞一下，所以可能要等一段时间。然后我看到DiffuQKT是你的工作，如果方便的话，麻烦提供一个model，要求实现一个forward(self, batch)就行，其中batch:dict[str, tensor]包含数据信息，例如batch["question_seq"]是习题id，shape是(bs, seq_len)，其它的还有batch["correctness_seq"]。我不知道DiffuQKT是否需要用到其它信息（如时间戳等）？

**Comment by ZhijieXiong** [comment]
另外一个问题就是我这边没有算力了，小数据集我可以自己租服务器测一下，大数据集如果你想用pyedmine测试一下的话，需要麻烦你提供一下算力

**Comment by ZhijieXiong** [comment]
没注意看，才发现ABQR和ReKT也是你的工作。太厉害了，ABQR和MIKT在性能上是实打实的SOTA，而且我觉得MIKT在可解释性上是目前做的最好的，佩服！

**Comment by lilstrawberry** [comment]
感谢

ReKT比较简单，放到kaggle上都能跑，不需要大内存，我觉得它的性能应该能吊打MIKT😂所以比较好奇哈哈哈哈。
我现在也没干科研了，已经当牛马了，只是单纯好奇ReKT的性能🤣DiffuQKT那个就算了😂

**Comment by lilstrawberry** [comment]
> 另外一个问题就是我这边没有算力了，小数据集我可以自己租服务器测一下，大数据集如果你想用pyedmine测试一下的话，需要麻烦你提供一下算力

KT代码比较少，不需要单独租服务器跑（除非有的需要Graph那种爆内存的）。在kaggle上白嫖每周30h的GPU就能跑了（验证手机号即可，注册多个账号就能并行跑）

**Comment by ZhijieXiong** [comment]
哈哈，好的，主要服务器用起来舒服一些。听你一说，我也好奇ReKT的性能了，我去复现一下。我感觉KT性能已经到瓶颈了，如果ReKT能远超MIKT，确实太nb了

**Comment by lilstrawberry** [comment]
刚刚瞅了一下论文的性能，吊打MIKT应该是吹牛🤣（不过稳稳压它一头倒应该没问题），ReKT本来把题目嵌入计算的好一点就能实现更强的性能（比如题目嵌入做个对比学习或者GCN啥的），但是当时为了留坑后面继续做没弄😂😂😂现在挖的坑没填🤑

**Comment by ZhijieXiong** [comment]
我跑了一些小数据集，可能是实验setting不一样，ReKT只在两个数据集上比MIKT强，不过还有几个大数据集在跑。ps：我用的参数是官方代码里的，即dim_emb=128,dim_state=128,dropout=0.4,learning_rate=0.002,grad_clip=15，没有调参，不过MIKT也没调参，所有数据集都用同一套参数，也是来自你的代码

**Comment by lilstrawberry** [comment]
🌚

**Comment by ZhijieXiong** [comment]
<img width="1277" height="312" alt="Image" src="https://github.com/user-attachments/assets/218d5a1d-2689-4bf0-bece-eda4d069d683" />

跑完了所有数据集，确实如果是和MIKT比的话，多数数据集上ReKT都更强

**Comment by lilstrawberry** [comment]
好的 感谢！
